[VDG] Take into account the Recipient label during pocket selection

soosr opened the pull request:
based on https://github.com/zkSNACKs/WalletWasabi/pull/7386

yahiheb commented on WalletWasabi.Fluent/ViewModels/Wallets/Send/LabelSelectionViewModel.cs:
```diff
@@ -32,18 +32,30 @@ public LabelSelectionViewModel(Money targetAmount)
 		var privateAndUnknownPockets = _allPockets.Except(knownPockets).ToArray();
 		var privateAndKnownPockets = _allPockets.Except(unknownPockets).ToArray();
+		var knownByRecipientPockets = knownPockets.Where(x => x.Labels.Any(label => recipient.Labels.Contains(label))).ToArray();
+		var onlyKnownByRecipientPockets = knownByRecipientPockets.Where(x => x.Labels.All(label => recipient.Labels.Contains(label))).ToArray();
```
Nit: var name should be `onlyKnownByRecipientPocket` since there is only one pocket.

soosr replied:
Yes, it could only contain one pocket, but that variable is an array, so that is why I went with the plural form.
I can make it to be only a `Pocket` object, but the method returns an array so it should be put back into an array when returning it. I found the current solution simple.

yahiheb replied:
I get it, no need I was just talking about the var name. Feel free to ignore.

yahiheb commented on WalletWasabi.Fluent/ViewModels/Wallets/Send/LabelSelectionViewModel.cs:
```diff
@@ -32,18 +32,30 @@ public LabelSelectionViewModel(Money targetAmount)
 			return new[] { _privatePocket };
 		}
 
+		if (onlyKnownByRecipientPockets.Sum(x => x.Amount) >= _targetAmount)
```
IMO this should be before selecting the private pocket if possible. We should get rid of non private coins whenever possible if that does not hurt our privacy.

If you use `onlyKnownByRecipientPockets` the recipient would learn nothing new about your wallet and the privacy level of the wallet would increase after that tx.

But by using private coins, the recipient would know more about our wallet after making this transaction (the coins that they already knew about + the change).

soosr replied:
I asked @molnard about what should be the best to do. 
And you are right in the case if the Labels are consistent and work as expected.
To not rely on this we agreed to go with the atom safe solution. Use the private funds first.

yahiheb replied:
But what is the point of the labels if we don't rely on them?

I guess everywhere in the software we rely on a correct usage by the user.

soosr replied:
@molnard Can we rediscuss this? I think @yahiheb has a point. We very much rely on the labels even on the Privacy Control screen.

soosr replied:
Ok, I will move this case to the first place.

soosr commented on WalletWasabi.Tests/UnitTests/UserInterfaceTest/PocketSelectionTests.cs:
```diff
@@ -582,12 +582,72 @@ public void AutoSelectAllPockets()
+		pockets.AddPocket(0.8M, out var pocket1, CoinPocketHelper.PrivateFundsText);
+		pockets.AddPocket(1.1M, out var pocket2, "Dan");
+		pockets.AddPocket(0.5M, out var pocket3, CoinPocketHelper.UnlabelledFundsText);
+		pockets.AddPocket(1.1M, out var pocket4, "David, Lucas");
```
`SmartLabel` is used in the `AddPocket` method so due to `,` separation, they are separated into two labels, so it doesn't change the result of the tests but you are definitely right. I fix them.

yahiheb commented on WalletWasabi.Tests/UnitTests/UserInterfaceTest/PocketSelectionTests.cs:
```diff
@@ -582,12 +582,72 @@ public void AutoSelectAllPockets()
+
+		var pockets = new List<Pocket>();
+		pockets.AddPocket(0.8M, out var pocket1, CoinPocketHelper.PrivateFundsText);
+		pockets.AddPocket(1.1M, out var pocket2, "Dan");
```
```suggestion
		pockets.AddPocket(1.1M, out var pocket2, "David");
```

By doing this the test does not pass but it should.

The problem is with this:
`var onlyKnownByRecipientPockets = knownByRecipientPockets.Where(x => x.Labels.All(label => recipient.Labels.Contains(label))).ToArray();`

We should only select coins that have the exact labels.

soosr replied:
Fixed, thanks

yahiheb commented on WalletWasabi.Fluent/ViewModels/Wallets/Send/LabelSelectionViewModel.cs:
```diff
@@ -32,18 +32,30 @@ public LabelSelectionViewModel(Money targetAmount)
 		var unknownPockets = NonPrivatePockets.Except(knownPockets).ToArray();
 		var privateAndUnknownPockets = _allPockets.Except(knownPockets).ToArray();
 		var privateAndKnownPockets = _allPockets.Except(unknownPockets).ToArray();
+		var knownByRecipientPockets = knownPockets.Where(pocket => pocket.Labels.Any(recipient.Contains)).ToArray();
```
AFAIU this is not correct.
If the recipient is `"David", "Lucas"` then we should not select coins that have only `"David"` or `"Lucas"`.

If the above is correct this is how I would correct the test:

```
public void AutoSelectKnownByRecipientPockets()
{
	var selection = new LabelSelectionViewModel(Money.Parse("1.0"));

	var pockets = new List<Pocket>();
	pockets.AddPocket(0.8M, out var pocket1, CoinPocketHelper.PrivateFundsText);
	pockets.AddPocket(1.1M, out var pocket2, "Dan");
	pockets.AddPocket(0.5M, out var pocket3, CoinPocketHelper.UnlabelledFundsText);
	pockets.AddPocket(1.1M, out var pocket4, "David", "Lucas");
	pockets.AddPocket(1.1M, out var pocket5, "David");
	pockets.AddPocket(1.1M, out var pocket6, "Lucas");
	pockets.AddPocket(1.1M, out var pocket7, "David", "Lucas", "Dan");

	selection.Reset(pockets.ToArray());

	var output = selection.AutoSelectPockets(new SmartLabel("David", "Lucas"));
	Assert.DoesNotContain(pocket1, output);
	Assert.DoesNotContain(pocket2, output);
	Assert.DoesNotContain(pocket3, output);
	Assert.Contains(pocket4, output);
	Assert.DoesNotContain(pocket5, output);
	Assert.DoesNotContain(pocket6, output);
	Assert.Contains(pocket7, output);
}
```

soosr replied:
Why? If we don't have coins with `David, Lucas` then it is better to use `David` or `Lucas` instead of a 3rd like `Dan`.

soosr replied:
The goal with this line:
```
var knownByRecipientPockets = knownPockets.Where(pocket => pocket.Labels.Any(recipient.Contains)).ToArray();
```
was to get all the pockets that contain ANY of the labels from the recipient list. What is the point of this? Better to use these pockets instead of those that have no relation at all with the recipients. Rename the variable to `knownByAnyRecipientPockets` maybe?

If there is a pocket with the exact match `onlyKnownByRecipientPockets` handles it.

What do you think about sorting the `knownByRecipientPockets` by the number of the labels that are found in the recipient?

So if the recipient is `David, Lucas`... and there are the pockets of `David`, `Lucas`, `David, Lucas, Dan` after sorting the `David, Lucas, Dan` will be in the first place so the transaction would use that.

Regarding you example:
```
public void AutoSelectKnownByRecipientPockets()
{
	var selection = new LabelSelectionViewModel(Money.Parse("1.0"));

	var pockets = new List<Pocket>();
	pockets.AddPocket(0.8M, out var pocket1, CoinPocketHelper.PrivateFundsText);
	pockets.AddPocket(1.1M, out var pocket2, "Dan");
	pockets.AddPocket(0.5M, out var pocket3, CoinPocketHelper.UnlabelledFundsText);
	pockets.AddPocket(1.1M, out var pocket4, "David", "Lucas");
	pockets.AddPocket(1.1M, out var pocket5, "David");
	pockets.AddPocket(1.1M, out var pocket6, "Lucas");
	pockets.AddPocket(1.1M, out var pocket7, "David", "Lucas", "Dan");

	selection.Reset(pockets.ToArray());

	var output = selection.AutoSelectPockets(new SmartLabel("David", "Lucas"));
	Assert.DoesNotContain(pocket1, output);
	Assert.DoesNotContain(pocket2, output);
	Assert.DoesNotContain(pocket3, output);
	Assert.Contains(pocket4, output);
	Assert.DoesNotContain(pocket5, output);
	Assert.DoesNotContain(pocket6, output);
	Assert.Contains(pocket7, output);
}
```

This should fail. Because the goal is `David, Lucas` and there is a pocket with the exact same labels which covers the `1 BTC` amount. Pocket 7 should not be included.

yahiheb replied:
> This should fail. Because the goal is David, Lucas and there is a pocket with the exact same labels which covers the 1 BTC amount. Pocket 7 should not be included.

That is true. This change should make it pass `pockets.AddPocket(0.9M, out var pocket4, "David", "Lucas");`

> The goal with this line:
> 
> var knownByRecipientPockets = knownPockets.Where(pocket => pocket.Labels.Any(recipient.Contains)).ToArray();
> was to get all the pockets that contain ANY of the labels from the recipient list. What is the point of this? Better to use these pockets instead of those that have no relation at all with the recipients. 

That line still can select pockets that have labels without any relation to the recipient, for example `David, Dan`.

Let's say the recipient is `David, Lucas` and you want to send 1 btc and you have the following coins:

coin1: 0.5 btc | "David"
coin2: 0.5 btc | "Lucas"
coin3: 1.0 btc | "David", "Lucas", "Dan"

Which coins are better to select/use?

The current implementation would select all 3 coins, what I am suggesting is we should only select the 3rd coin.

Even if you somehow select only coin1 and coin2 you are leaking some information to David AND to Lucas, while if you use coin3 you are only leaking some information to Dan.

soosr replied:
> That line still can select pockets that have labels without any relation to the recipient, for example David, Dan.
 
 I am not sure I understand this. If the Recipient is `David, Lucas` and it selects a pocket with `David, Dan` I clearly see there is a relation. Both contain `David`. So this is an acceptable selection.
 
>Let's say the recipient is David, Lucas and you want to send 1 btc and you have the following coins:
> 
> coin1: 0.5 btc | "David"
> coin2: 0.5 btc | "Lucas"
> coin3: 1.0 btc | "David", "Lucas", "Dan"
> 
> Which coins are better to select/use?
>
> The current implementation would select all 3 coins, what I am suggesting is we should only select the 3rd coin.

And yes you are right. In this example David and Lucas is an acceptable selection but `David, Lucas, Dan` is the best selection.

My proposal about the sorting `What do you think about sorting the knownByRecipientPockets by the number of the labels that are found in the recipient?` would exactly do what you want. It would select the pocket that contains the most labels from the recipient labels.

My only concern is this is starting to be more than a UI concept. Initially, we just selected between two pockets. Private / non private, then we distinguished the known / unknown pockets. Which still fits into the UI since the library code has no idea about them. But right now we are trying to select the exact and best coins which normally would be the SmartCoinSelector's task IMO. Maybe I am wrong but it seems we started to mix business logic with UI.

soosr replied:
Added additional logic.

But in my honest opinion, this is surely mixing business logic with UI and I am sure there are still edge cases when it does not find the best pockets... (or at least I don't know which one should be the best)... I start to feel why Lucas stopped it [here](https://github.com/zkSNACKs/WalletWasabi/pull/2591)

yahiheb replied:
I agree this is mixing mixing business logic and UI.

> I start to feel why Lucas stopped it https://github.com/zkSNACKs/WalletWasabi/pull/2591

I get that, but that was not as much important as it is now, because we are doing the selection instead of the user.

My goal with this is to have the minimum obvious use cases, and later on obviously we can improve it more.

soosr replied:
I improved the logic and I believe it is working correctly now... at least I believe, would be nice to have tests for the most common cases?

yahiheb reviewed (changes requested):
To rely less on a correct usage by the users we can make the pocket selection smarter, the labels are checked with case insensitive, so coins with "Dan" or "dan" are considered one pocket.

soosr commented:
> To rely less on a correct usage by the users we can make the pocket selection smarter, the labels are checked with case insensitive, so coins with "Dan" or "dan" are considered one pocket.

This should never happen. The user should not be able to enter `dan` if they already entered `Dan` before. If it can happen that it is a bug and the bug should be fixed instead of handling this case.

soosr commented:
> > This should never happen. The user should not be able to enter `dan` if they already entered `Dan` before. If it can happen that it is a bug and the bug should be fixed instead of handling this case.
> 
> It was possible with WW1. What about if the user types `DAn`?

Same. If you have `Dan` entered, you have no chance to enter any other capitalized version of that text. Although this is just a UI concept and if someone hacks the wallet file or the transaction.dat file the case can happen.

yahiheb commented:
> Same. If you have `Dan` entered, you have no chance to enter any other capitalized version of that text. Although this is just a UI concept and if someone hacks the wallet file or the transaction.dat file the case can happen.

But what we currently have is not consistent at all. If I have `Dan` with capital `D` then I can't enter any other label that starts with lower `d` like `david` and vice versa. Check #6354.

> It was possible with WW1.

What about this?

MaxHillebrand commented:
@yahiheb it might be because you created a `dan` and `Dan` label before this PR, and now it's like if you would be editing the wallet file labels in this PR.

Try a new wallet with no labels, and then it should work as intended.

yahiheb commented:
> @yahiheb it might be because you created a `dan` and `Dan` label before this PR, and now it's like if you would be editing the wallet file labels in this PR.

No. Please check the linked issue to know what I am talking about.

> Try a new wallet with no labels, and then it should work as intended.

Labels have nothing to do with the wallet being new or not.

yahiheb commented on WalletWasabi.Fluent/ViewModels/Wallets/Send/LabelSelectionViewModel.cs:
```diff
@@ -32,18 +33,30 @@ public LabelSelectionViewModel(Money targetAmount)
 			return new[] { _privatePocket };
 		}
 
+		if (GetBestKnownByRecipientPockets(knownByRecipientPockets, _targetAmount, recipient) is { } pockets)
```
We should check the sum of the amounts here:

```suggestion
		if (GetBestKnownByRecipientPockets(knownByRecipientPockets, _targetAmount, recipient) is { } pockets && pockets.Sum(x => x.Amount) >= _targetAmount)
```

Otherwise this test wouldn't pass:

```
public void MyTest()
{
	var selection = new LabelSelectionViewModel(Money.Parse("1.0"));

	var pockets = new List<Pocket>();
	pockets.AddPocket(0.8M, out var pocket1, CoinPocketHelper.PrivateFundsText);
	pockets.AddPocket(1.1M, out var pocket2, "Dan");
	pockets.AddPocket(0.5M, out var pocket3, CoinPocketHelper.UnlabelledFundsText);
	pockets.AddPocket(0.1M, out var pocket4, "David", "Lucas");
	pockets.AddPocket(0.15M, out var pocket5, "David");
	pockets.AddPocket(0.2M, out var pocket6, "Lucas");
	pockets.AddPocket(0.3M, out var pocket7, "David", "Lucas", "Dan");

	selection.Reset(pockets.ToArray());

	var output = selection.AutoSelectPockets(new SmartLabel("David", "Lucas"));
	Assert.DoesNotContain(pocket1, output);
	Assert.Contains(pocket2, output);
	Assert.DoesNotContain(pocket3, output);
	Assert.Contains(pocket4, output);
	Assert.Contains(pocket5, output);
	Assert.Contains(pocket6, output);
	Assert.Contains(pocket7, output);
}
```

If you agree this test is a better replacement to the `AutoSelectOnlyUnknownPocket` test.

soosr replied:
> We should check the sum of the amounts here:

Fixed in another way, I made a mistake when I checked if the search was successful or not.

> If you agree this test is a better replacement to the AutoSelectOnlyUnknownPocket test.

You mean `AutoSelectOnlyKnownPockets`?

yahiheb commented:
@soosr I pushed https://github.com/zkSNACKs/WalletWasabi/pull/7443/commits/aeef2588dbbf2137515a614fe28f75dff12c74f7 to reflect the changes by https://github.com/zkSNACKs/WalletWasabi/pull/7443/commits/4a16342d217f464579fa954cd85fd2cb2d27b650

Also IMO we should just rename `AutoSelectOnlyKnownByAllMultipleRecipientPockets` to `AutoSelectOnlyKnownByRecipientPockets` and remove the original `AutoSelectOnlyKnownByRecipientPockets` because they cover the same case.

soosr commented:
> > It was possible with WW1.
> 
> What about this?

Ok, since it was possible in WW1, probably there are a few wallets so agree. I will make it case insensitive.

soosr commented:
I addressed all review comments, ready for testing. Although I am not sure if this is worth it. To heavy business logic in the UI for the sake of edge cases. Anyway, let's see how it works and decide.

soosr commented on WalletWasabi.Fluent/ViewModels/Wallets/Send/LabelSelectionViewModel.cs:
```diff
@@ -72,6 +90,53 @@ public Pocket[] AutoSelectPockets()
+			}
+		}
+
+		foreach (var p in pockets.OrderBy(x => x.Amount).ToImmutableArray())
+		{
+			if (pockets.Sum(x => x.Amount) - p.Amount >= targetAmount)
+			{
+				pockets.Remove(p);
+			}
+			else
+			{
+				break;
+			}
+		}
```
In the `privacyRankedPockets` you have pockets [A,B,C,D,E].
You start to select pockets from the first to the last from that list until they cover the target amount.
Let's say A, B, and C cover it. But it can happen that B and C are enough and A is unnecessary. So here I order the pockets by amounts `pockets.OrderBy(x => x.Amount).ToImmutableArray()` and check if the remaining `pockets` without `p` does still cover the target amount. If so then `p` is not necessary.
So continue the example... pocket A is the best from the privacy perspective but it is not necessary, so here it will be removed. 

And the result is B and C. And we also know that D and E are worse from the privacy perspective so we don't have to deal with them. 

I will add test to this case.

yahiheb replied:
> So here I order the pockets by amounts `pockets.OrderBy(x => x.Amount).ToImmutableArray()` 

I think we should order by amounts AND also by number of labels, so if the amounts of A and B are equal then we should take A and remove B.

`foreach (var p in bestPockets.OrderBy(x => x.Amount).ThenByDescending(x => x.Labels.Count()).ToImmutableArray())`

If you agree with that then we can edit `AutoSelectRequiredKnownByRecipientPockets` as follow:

```
public void AutoSelectRequiredKnownByRecipientPockets()
{
	var selection = new LabelSelectionViewModel(Money.Parse("1.0"));

	var pockets = new List<Pocket>();
	pockets.AddPocket(0.8M, out var pocket1, CoinPocketHelper.PrivateFundsText);
	pockets.AddPocket(1.1M, out var pocket2, "Dan");
	pockets.AddPocket(1.5M, out var pocket3, CoinPocketHelper.UnlabelledFundsText);
	pockets.AddPocket(0.4M, out var pocket4, "David");
	pockets.AddPocket(0.4M, out var pocket5, "David", "Max");
	pockets.AddPocket(0.6M, out var pocket6, "David", "Lucas", "Dan");

	selection.Reset(pockets.ToArray());

	var output = selection.AutoSelectPockets("David");
	Assert.DoesNotContain(pocket1, output);
	Assert.DoesNotContain(pocket2, output);
	Assert.DoesNotContain(pocket3, output);
	Assert.Contains(pocket4, output);
	Assert.DoesNotContain(pocket5, output);
	Assert.Contains(pocket6, output);
}
```

soosr replied:
Done

soosr commented on WalletWasabi.Fluent/ViewModels/Wallets/Send/LabelSelectionViewModel.cs:
```diff
@@ -32,18 +33,35 @@ public LabelSelectionViewModel(Money targetAmount)
 			return new[] { _privatePocket };
 		}
 
+		if (GetBestKnownByRecipientPockets(knownByRecipientPockets, _targetAmount, recipient) is { } pockets)
```
I don't prefer this, because if someone used the `GetBestKnownByRecipientPockets` method without checking the amount before calling it, the method could return pockets that do not cover `_targetAmount`. For safety reasons, the amount check should be inside of the method.

I modified it. It does not do unnecessary calculations and I removed the nullable.

yahiheb commented on WalletWasabi.Fluent/ViewModels/Wallets/Send/LabelSelectionViewModel.cs:
```diff
@@ -72,6 +91,59 @@ public Pocket[] AutoSelectPockets()
+			}
+		}
+
+		foreach (var p in bestPockets.OrderBy(x => x.Amount).ToImmutableArray())
```
We should also order by number of labels here. See https://github.com/zkSNACKs/WalletWasabi/pull/7443#discussion_r823354315

```suggestion
		foreach (var p in bestPockets.OrderBy(x => x.Amount).ThenByDescending(x => x.Labels.Count()).ToImmutableArray())
```

MaxHillebrand reviewed (approved):
tACK [37f3c0a](https://github.com/zkSNACKs/WalletWasabi/pull/7443/commits/37f3c0aee0dac03378b235159165396be8b1fc4c)

yahiheb commented:
Even though `AutoSelectPockets()` method selected only 1 pocket `onlyKnownByRecipientPockets`, I still see the other pockets in the control privacy dialog.

![sss](https://user-images.githubusercontent.com/52379387/157689388-217d0614-c455-4acc-af9d-55b15ac9e869.PNG)

This should not be the case IMO, it is like saying to the user "hey we selected the best private pocket for your transaction but if you wanna screw it go ahead".

Why do we not do the same when we select private coins then? and give the user the possibly to use less private pockets.

soosr commented:
> Even though `AutoSelectPockets()` method selected only 1 pocket `onlyKnownByRecipientPockets`, I still see the other pockets in the control privacy dialog.
> 
> ![sss](https://user-images.githubusercontent.com/52379387/157689388-217d0614-c455-4acc-af9d-55b15ac9e869.PNG)
> 
> This should not be the case IMO, it is like saying to the user "hey we selected the best private pocket for your transaction but if you wanna screw it go ahead".
> 
> Why do we not do the same when we select private coins then? and give the user the possibly to use less private pockets.
> 
> Edit: this is maybe not related to this PR but anyway.

I have a branch where I am working on this to do not show the "Control your privacy" dialog and suggestion when there is no better selection. Deal with this there.

yahiheb reviewed (approved):
@soosr Very good job.

This is a good progress in the right direction to have a solid pocket/coin selection, specially since we don't have coin control and we are doing the selection instead of the user.

tACK

MaxHillebrand commented:
This looks really solid.

Questuon: how are unconfirmed coins considered?

soosr commented:
> This looks really solid.
> 
> Question: how are unconfirmed coins considered?

What do you mean? This PR doesn't deal with unconfirmed coins.

MaxHillebrand commented:
I receive from `Alice`, the transaction is unconfirmed. 

Now I want to send to `Alice`.

Does it consider the unconfirmed coin, and what is the fee strategy for the child transaction?

Or does it only select out of the confirmed coins?

More of a general question unrelated to this PR, do unconfirmed coins get considered for auto coinselection in payments [they're not for coinjoins]?

It's just interesting how this applies to this PR specifically.

yahiheb commented:
> Question: how are unconfirmed coins considered?

AFAIK we don't have any special treatment for unconfirmed coins when sending, and this PR does not change that.

> I receive from `Alice`, the transaction is unconfirmed.
> 
> Now I want to send to `Alice`.
> 
> Does it consider the unconfirmed coin,

Yes.

> and what is the fee strategy for the child transaction?

IDK

> More of a general question unrelated to this PR, do unconfirmed coins get considered for auto coinselection in payments [they're not for coinjoins]?

Again AFAIK yes, but here we are doing the pocket selection.

MaxHillebrand commented:
Closes: #2538

danwalmsley reviewed (approved):
LGTM
